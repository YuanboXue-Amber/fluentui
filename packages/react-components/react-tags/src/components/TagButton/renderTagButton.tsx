import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { TagButtonState, TagButtonSlots } from './TagButton.types';

/**
 * Render the final JSX of TagButton
 */
export const renderTagButton_unstable = (state: TagButtonState) => {
  const { slots, slotProps } = getSlots<TagButtonSlots>(state);

  // TODO Add additional slots in the appropriate place
  return (
    <slots.root {...slotProps.root}>
      <button className="content">
        <div className="persona">
          <img
            height="32"
            width="32"
            src="https://res-1.cdn.office.net/files/fabric-cdn-prod_20221209.001/office-ui-fabric-react-assets/persona-male.png"
          />
        </div>
        <div className="icon">i</div>
        <div className="primary">Primary Text</div>
        <div className="secondary">Secondary Text</div>
      </button>
      <div className="dismiss">
        <button>x</button>
      </div>
    </slots.root>
  );
};
