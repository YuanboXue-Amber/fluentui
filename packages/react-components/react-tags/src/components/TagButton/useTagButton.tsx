import * as React from 'react';
import { resolveShorthand, useId } from '@fluentui/react-utilities';
import type { TagButtonProps, TagButtonState } from './TagButton.types';
import { useTag_unstable } from '../Tag/index';
import { useTagGroupContext_unstable } from '../../contexts/TagGroupContext';
import type { ARIAButtonElement, ARIAButtonElementIntersection } from '@fluentui/react-aria';

/**
 * Create the state required to render TagButton.
 *
 * The returned state can be modified with hooks such as useTagButtonStyles_unstable,
 * before being passed to renderTagButton_unstable.
 *
 * @param props - props from this instance of TagButton
 * @param ref - reference to root HTMLElement of TagButton
 */
export const useTagButton_unstable = (
  props: TagButtonProps,
  ref: React.Ref<ARIAButtonElement<'div'>>,
): TagButtonState => {
  const toggleCheckedItems = useTagGroupContext_unstable(context => context.toggleCheckedItems);

  const id = useId('fui-TagButton-', props.id);

  const contextChecked = useTagGroupContext_unstable(context => {
    const checkedItems = context.checkedItems || [];
    return checkedItems.indexOf(id) !== -1;
  });
  const { checked = contextChecked } = props;

  // TODO to select on keydown, useARIAButtonShorthand, but that would mean content is 'button' and we'll have trouble re-use useTag hook
  const content = resolveShorthand(props.content, {
    required: true,
    defaultProps: {
      tabIndex: 0,
      'aria-checked': checked,
      // TODO aria attributes
      onClick: (e: React.MouseEvent<ARIAButtonElementIntersection<'div'>>) => {
        toggleCheckedItems?.(e, id, checked);
        props.onClick?.(e);
      },
    },
  });

  return useTag_unstable({ ...props, content, checked }, ref);
};
