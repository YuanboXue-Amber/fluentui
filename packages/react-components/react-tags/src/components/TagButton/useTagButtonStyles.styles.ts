import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { TagButtonSlots, TagButtonState } from './TagButton.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { createCustomFocusIndicatorStyle } from '@fluentui/react-tabster';
import { tokens } from '@fluentui/react-theme';
import { useTagBaseStyles } from '../Tag/index';
import { iconFilledClassName, iconRegularClassName } from '@fluentui/react-icons';

export const tagButtonClassNames: SlotClassNames<TagButtonSlots> = {
  root: 'fui-TagButton',
  content: 'fui-TagButton__content',
  media: 'fui-TagButton__media',
  icon: 'fui-TagButton__icon',
  primaryText: 'fui-TagButton__primaryText',
  secondaryText: 'fui-TagButton__secondaryText',
  dismissButton: 'fui-TagButton__dismissButton',
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  content: {
    position: 'relative',
    ...createCustomFocusIndicatorStyle({
      ...shorthands.borderColor(tokens.colorTransparentStroke),
      ...shorthands.borderRadius(tokens.borderRadiusMedium),
      ...shorthands.outline(tokens.strokeWidthThick, 'solid', tokens.colorTransparentStroke),
      boxShadow: `
      ${tokens.shadow4},
      0 0 0 2px ${tokens.colorStrokeFocus2}
    `,
      zIndex: 1,
    }),
  },

  circularContent: createCustomFocusIndicatorStyle(shorthands.borderRadius(tokens.borderRadiusCircular)),
  dismissableContent: {
    ...createCustomFocusIndicatorStyle({
      borderTopRightRadius: tokens.borderRadiusNone,
      borderBottomRightRadius: tokens.borderRadiusNone,
    }),
  },

  dismissButton: {
    ...shorthands.padding('0px', '6px'),
    ...shorthands.borderLeft(tokens.strokeWidthThin, 'solid', tokens.colorNeutralStroke1),
    borderTopLeftRadius: tokens.borderRadiusNone,
    borderBottomLeftRadius: tokens.borderRadiusNone,
    ...createCustomFocusIndicatorStyle({
      borderTopLeftRadius: tokens.borderRadiusNone,
      borderBottomLeftRadius: tokens.borderRadiusNone,
    }),
  },
  dismissButtonCircular: createCustomFocusIndicatorStyle({
    borderTopRightRadius: tokens.borderRadiusCircular,
    borderBottomRightRadius: tokens.borderRadiusCircular,
  }),

  // TODO add additional classes for fill/outline appearance, different sizes, and state
});

const useCheckedStyles = makeStyles({
  root: {
    backgroundColor: tokens.colorBrandBackground2,
    // TODO put border in base style as well to prevent reflow
    ...shorthands.border(tokens.strokeWidthThin, 'solid', tokens.colorBrandStroke2),
    color: tokens.colorBrandForeground2,
  },
  content: {
    [`& .${iconFilledClassName}`]: {
      display: 'inline',
    },
    [`& .${iconRegularClassName}`]: {
      display: 'none',
    },
  },
});

/**
 * Apply styling to the TagButton slots based on the state
 */
export const useTagButtonStyles_unstable = (state: TagButtonState): TagButtonState => {
  const baseStyles = useTagBaseStyles();
  const styles = useStyles();
  const checkedStyles = useCheckedStyles();

  state.root.className = mergeClasses(
    tagButtonClassNames.root,
    baseStyles.root,
    state.shape === 'circular' && baseStyles.rootCircular,
    state.checked && checkedStyles.root,
    state.root.className,
  );
  if (state.content) {
    state.content.className = mergeClasses(
      tagButtonClassNames.content,
      baseStyles.content,
      !state.media && !state.icon && baseStyles.textOnlyContent,

      styles.content,
      state.shape === 'circular' && styles.circularContent,
      state.dismissButton && styles.dismissableContent,

      state.checked && checkedStyles.content,

      state.content.className,
    );
  }

  if (state.media) {
    state.media.className = mergeClasses(tagButtonClassNames.media, baseStyles.media, state.media.className);
  }
  if (state.icon) {
    state.icon.className = mergeClasses(tagButtonClassNames.icon, baseStyles.icon, state.icon.className);
  }
  if (state.primaryText) {
    state.primaryText.className = mergeClasses(
      tagButtonClassNames.primaryText,
      baseStyles.primaryText,
      state.secondaryText && baseStyles.primaryTextWithSecondaryText,
      state.primaryText.className,
    );
  }
  if (state.secondaryText) {
    state.secondaryText.className = mergeClasses(
      tagButtonClassNames.secondaryText,
      baseStyles.secondaryText,
      state.secondaryText.className,
    );
  }
  if (state.dismissButton) {
    state.dismissButton.className = mergeClasses(
      tagButtonClassNames.dismissButton,
      baseStyles.resetButton,
      baseStyles.dismissButton,

      styles.dismissButton,
      state.shape === 'circular' && styles.dismissButtonCircular,
      state.dismissButton.className,
    );
  }

  return state;
};
