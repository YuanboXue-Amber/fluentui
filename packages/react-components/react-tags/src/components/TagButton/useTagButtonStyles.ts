import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { TagButtonSlots, TagButtonState } from './TagButton.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { createCustomFocusIndicatorStyle } from '../../../../react-tabster/src/index';
import { tokens, typographyStyles } from '../../../../react-theme/src/index';

export const tagButtonClassNames: SlotClassNames<TagButtonSlots> = {
  root: 'fui-TagButton',
  content: 'fui-TagButton__content',
  media: 'fui-TagButton__media',
  icon: 'fui-TagButton__icon',
  primaryText: 'fui-TagButton__primaryText',
  secondaryText: 'fui-TagButton__secondaryText',
  dismissButton: 'fui-TagButton__dismissButton',
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    display: 'inline-flex',
    height: '32px',
    width: 'fit-content',
    ...shorthands.borderRadius(tokens.borderRadiusMedium),

    ...shorthands.outline('1px', 'solid', 'red'),

    backgroundColor: tokens.colorNeutralBackground3,
    color: tokens.colorNeutralForeground2,
  },
  rootCircular: {
    ...shorthands.borderRadius(tokens.borderRadiusCircular),
  },

  media: {
    ...shorthands.gridArea('media'),
    alignSelf: 'center',
    paddingLeft: tokens.spacingHorizontalXXS,
    paddingRight: tokens.spacingHorizontalS,
  },

  content: {
    display: 'inline-grid',
    gridTemplateRows: '1fr auto auto 1fr',
    gridTemplateAreas: `
    "media icon .        "
    "media icon primary  "
    "media icon secondary"
    "media icon .        "
    `,
    paddingRight: tokens.spacingHorizontalS,

    position: 'relative',
    ...createCustomFocusIndicatorStyle({
      ...shorthands.borderColor(tokens.colorTransparentStroke),
      ...shorthands.borderRadius(tokens.borderRadiusMedium),
      ...shorthands.outline(tokens.strokeWidthThick, 'solid', tokens.colorTransparentStroke),
      boxShadow: `
      ${tokens.shadow4},
      0 0 0 2px ${tokens.colorStrokeFocus2}
    `,
      zIndex: 1,
    }),
  },
  textOnlyContent: {
    paddingLeft: tokens.spacingHorizontalS,
  },
  circularContent: createCustomFocusIndicatorStyle(shorthands.borderRadius(tokens.borderRadiusCircular)),
  dismissableContent: createCustomFocusIndicatorStyle({
    borderTopRightRadius: tokens.borderRadiusNone,
    borderBottomRightRadius: tokens.borderRadiusNone,
  }),

  icon: {
    display: 'flex',
    alignSelf: 'center',
    ...shorthands.gridArea('icon'),
    paddingLeft: '6px',
    paddingRight: '2px',
  },
  primaryText: {
    gridColumnStart: 'primary',
    gridRowStart: 'primary',
    gridRowEnd: 'secondary',
    ...typographyStyles.body1,
    paddingLeft: tokens.spacingHorizontalXXS,
    paddingRight: tokens.spacingHorizontalXXS,
  },
  primaryTextWithSecondaryText: {
    ...shorthands.gridArea('primary'),
    ...typographyStyles.caption1,
  },
  secondaryText: {
    ...shorthands.gridArea('secondary'),
    paddingLeft: tokens.spacingHorizontalXXS,
    paddingRight: tokens.spacingHorizontalXXS,
    ...typographyStyles.caption2,
  },

  dismissButton: {
    minWidth: '20px',
    ...shorthands.padding('0px', '6px'),
    borderLeftColor: tokens.colorNeutralStroke1,
    borderTopLeftRadius: tokens.borderRadiusNone,
    borderBottomLeftRadius: tokens.borderRadiusNone,

    ...createCustomFocusIndicatorStyle({
      borderTopLeftRadius: tokens.borderRadiusNone,
      borderBottomLeftRadius: tokens.borderRadiusNone,
    }),
  },
  dismissButtonCircular: createCustomFocusIndicatorStyle({
    borderTopRightRadius: tokens.borderRadiusCircular,
    borderBottomRightRadius: tokens.borderRadiusCircular,
  }),

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the TagButton slots based on the state
 */
export const useTagButtonStyles_unstable = (state: TagButtonState): TagButtonState => {
  const styles = useStyles();

  state.root.className = mergeClasses(
    tagButtonClassNames.root,
    styles.root,
    state.shape === 'circular' && styles.rootCircular,
    state.root.className,
  );
  if (state.content) {
    state.content.className = mergeClasses(
      tagButtonClassNames.content,
      styles.content,
      !state.media && !state.icon && styles.textOnlyContent,

      state.shape === 'circular' && styles.circularContent,
      state.dismissButton && styles.dismissableContent,

      state.content.className,
    );
  }

  if (state.media) {
    state.media.className = mergeClasses(tagButtonClassNames.media, styles.media, state.media.className);
  }
  if (state.icon) {
    state.icon.className = mergeClasses(tagButtonClassNames.icon, styles.icon, state.icon.className);
  }
  if (state.primaryText) {
    state.primaryText.className = mergeClasses(
      tagButtonClassNames.primaryText,
      styles.primaryText,
      state.secondaryText && styles.primaryTextWithSecondaryText,
      state.primaryText.className,
    );
  }
  if (state.secondaryText) {
    state.secondaryText.className = mergeClasses(
      tagButtonClassNames.secondaryText,
      styles.secondaryText,
      state.secondaryText.className,
    );
  }
  if (state.dismissButton) {
    state.dismissButton.className = mergeClasses(
      tagButtonClassNames.dismissButton,
      styles.dismissButton,
      state.shape === 'circular' && styles.dismissButtonCircular,
      state.dismissButton.className,
    );
  }

  return state;
};
