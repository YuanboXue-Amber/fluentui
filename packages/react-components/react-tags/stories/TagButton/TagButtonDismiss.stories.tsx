import * as React from 'react';
import { Avatar } from '@fluentui/react-avatar';
import { Calendar3Day20Regular } from '@fluentui/react-icons';

import { TagButton } from '@fluentui/react-tags';

export const Dismiss = () => (
  <div style={{ display: 'flex', flexDirection: 'column', gap: '10px' }}>
    <TagButton dismissable>Primary text</TagButton>
    <TagButton dismissable icon={<Calendar3Day20Regular />}>
      Primary text
    </TagButton>
    <TagButton
      dismissable
      media={<Avatar name="Katri Athokas" badge={{ status: 'busy' }} />}
      secondaryText="Secondary text"
    >
      Primary text
    </TagButton>
  </div>
);

Dismiss.storyName = 'Dismiss';
Dismiss.parameters = {
  docs: {
    description: {
      story: 'A TagButton can have a button that dismisses it',
    },
  },
};
