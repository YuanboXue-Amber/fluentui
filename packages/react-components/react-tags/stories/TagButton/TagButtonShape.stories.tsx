import * as React from 'react';
import { Avatar } from '@fluentui/react-components';
import { Calendar3Day20Regular } from '@fluentui/react-icons';

import { TagButton } from '@fluentui/react-tags';

export const Shape = () => (
  <div style={{ display: 'flex', gap: '10px', flexDirection: 'column' }}>
    <div style={{ display: 'flex', gap: '10px' }}>
      <TagButton media={<Avatar name="Katri Athokas" badge={{ status: 'busy' }} />}>Rounded</TagButton>
      <TagButton shape="circular" media={<Avatar name="Katri Athokas" badge={{ status: 'busy' }} />}>
        Circular
      </TagButton>
    </div>
    <div style={{ display: 'flex', gap: '10px' }}>
      <TagButton dismissable icon={<Calendar3Day20Regular />} secondaryText="Secondary text">
        Rounded
      </TagButton>
      <TagButton shape="circular" dismissable icon={<Calendar3Day20Regular />} secondaryText="Secondary text">
        Circular
      </TagButton>
    </div>
  </div>
);

Shape.storyName = 'Shape';
Shape.parameters = {
  docs: {
    description: {
      story: 'A TagButton can be rounded or circular,',
    },
  },
};
