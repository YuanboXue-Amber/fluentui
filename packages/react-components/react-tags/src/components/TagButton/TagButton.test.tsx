import * as React from 'react';
import { TagButton } from './TagButton';
import { isConformant } from '../../testing/isConformant';
import { render, fireEvent } from '@testing-library/react';

const requiredProps = {
  media: 'media',
  icon: 'i',
  primaryText: 'Primary text',
  secondaryText: 'Secondary text',
  dismissible: true,
};

describe('TagButton', () => {
  isConformant({
    Component: TagButton,
    displayName: 'TagButton',
    requiredProps,
    disabledTests: [
      // onDismiss: A second (data) argument is not needed
      'consistent-callback-args',
    ],
  });

  it('should invoke onDismiss on dismiss button click', () => {
    const onDismiss = jest.fn();
    const { getByTestId } = render(
      <TagButton dismissible onDismiss={onDismiss} dismissButton={<div data-testid="foo" />}>
        Tag
      </TagButton>,
    );

    fireEvent.click(getByTestId('foo'));

    expect(onDismiss).toHaveBeenCalledTimes(1);
  });

  it('should invoke onDismiss on dismiss button delete keyDown', () => {
    const onDismiss = jest.fn();
    const { getByTestId } = render(
      <TagButton dismissible onDismiss={onDismiss} dismissButton={<div data-testid="foo" />}>
        Tag
      </TagButton>,
    );

    fireEvent.keyDown(getByTestId('foo'), { key: 'Delete' });

    expect(onDismiss).toHaveBeenCalledTimes(1);
  });
});
