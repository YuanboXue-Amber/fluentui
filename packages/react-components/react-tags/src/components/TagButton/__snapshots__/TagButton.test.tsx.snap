// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagButton renders a default state 1`] = `
<div>
  <div
    class="fui-TagButton"
  >
    <div
      class="fui-TagButton__content"
      tabindex="0"
    >
      <span
        class="fui-TagButton__media"
      >
        media
      </span>
      <span
        class="fui-TagButton__icon"
      >
        i
      </span>
      <span
        class="fui-TagButton__primaryText"
      >
        Default TagButton
      </span>
      <span
        class="fui-TagButton__secondaryText"
      >
        Secondary text
      </span>
    </div>
    <button
      class="fui-Button fui-TagButton__dismissButton"
      type="button"
    >
      <span
        class="fui-Button__icon"
      >
        <svg
          aria-hidden="true"
          class="fui-Icon-filled"
          fill="currentColor"
          height="1em"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3.9 4.05.07-.08a.75.75 0 0 1 .98-.07l.08.07L10 8.94l4.97-4.97a.75.75 0 0 1 .98-.07l.08.07c.27.27.3.68.07.98l-.07.08L11.06 10l4.97 4.97c.27.27.3.68.07.98l-.07.08a.75.75 0 0 1-.98.07l-.08-.07L10 11.06l-4.97 4.97a.75.75 0 0 1-.98.07l-.08-.07a.75.75 0 0 1-.07-.98l.07-.08L8.94 10 3.97 5.03a.75.75 0 0 1-.07-.98l.07-.08-.07.08Z"
            fill="currentColor"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="fui-Icon-regular"
          fill="currentColor"
          height="1em"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
            fill="currentColor"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
